### PR TITLE
Adding custom domain implementaion

### PR DESCRIPTION
**Pull request by lmanohara**

### CODE REVIEW

File: modules/components/org.wso2.appcloud.provisioning.runtime/src/main/java/org/wso2/appcloud/provisioning/runtime/KubernetesRuntimeProvisioningService.java [hunk 1]
```diff
@@ -805,8 +805,7 @@ public boolean createDeploymentUrl(String environmentUrl) throws RuntimeProvisio
                     .withKind(KubernetesPovisioningConstants.KIND_INGRESS)
                     .withNewMetadata()
                     .withName(
-                            KubernetesProvisioningUtils.createIngressMetaName(applicationContext, environmentUrl,
-                                    service.getMetadata().getName()))
+                            KubernetesProvisioningUtils.createIngressMetaName(environmentUrl))
```
Comment (pirinthapan) [inline modules/components/org.wso2.appcloud.provisioning.runtime/src/main/java/org/wso2/appcloud/provisioning/runtime/KubernetesRuntimeProvisioningService.java]: The line 808 can be moved to 807

Comment (lmanohara) [inline modules/components/org.wso2.appcloud.provisioning.runtime/src/main/java/org/wso2/appcloud/provisioning/runtime/KubernetesRuntimeProvisioningService.java]: This is fixed in following commit.
https://github.com/wso2/app-cloud/pull/167/commits/e0097c953bc01c32d65e4e6a120e23b356025d4f

File: modules/components/org.wso2.appcloud.provisioning.runtime/src/main/java/org/wso2/appcloud/provisioning/runtime/Utils/KubernetesProvisioningUtils.java [hunk 2]
```diff
@@ -124,15 +124,12 @@ public static ServiceList getServices(ApplicationContext applicationContext){
...
      * @param domain
-     * @param serviceName
      * @return generated unique name for ingress (appName-appVersion-service)
      */
-    public static String createIngressMetaName(ApplicationContext applicationContext, String domain, String serviceName){
+    public static String createIngressMetaName(String domain){
 
-        return (applicationContext.getId() + "-" + applicationContext.getVersion() + "-" + domain + "-" + serviceName)
-                .replace(".","-").toLowerCase();
+        return (domain).replace(".","-").toLowerCase();
```
Comment (danushkaf) [inline modules/components/org.wso2.appcloud.provisioning.runtime/src/main/java/org/wso2/appcloud/provisioning/runtime/Utils/KubernetesProvisioningUtils.java]: What is "." and what is "-"?. Make them constants with meaningful names.

File: modules/jaggeryapps/appmgt/src/modules/application/application.jag [hunk 3]
```diff
@@ -283,6 +285,17 @@ var deployApplication = function(applicationName, applicationRevision, deploymen
     }
 };
 
+var constructAppLaunchURL = function(applicationName, applicationRevision){
+    var urlPrefix = modManager.getTenantDomain()
+            + "-" + applicationName + "-" + applicationRevision;
+    var appLaunchURL = urlPrefix.replace(/\./g, '-')
+            +  "." + modManager.getPropertyValue("ApplicationLaunchBaseURL");
```
Comment (danushkaf) [inline modules/jaggeryapps/appmgt/src/modules/application/application.jag]: comment explaining why and what we are replacing. better to make '-' a constant.

File: modules/jaggeryapps/appmgt/src/modules/application/application.jag [hunk 4]
```diff
@@ -791,17 +804,12 @@ var getAllRevisionsOfApplication =function(appName){
...
-    var deploymentPath = KubernetesProvisioningUtils.getDeploymentPath(applicationCtx);
-    ///
     var urlScheme = CloudUtil.getPropertyValue("DeploymentUrlScheme");
-    var baseURL = urlScheme + "://" + CloudUtil.getPropertyValue("ApplicationLaunchBaseURL");
+    var baseURL = constructAppLaunchURL(applicationName, applicationRevision);
+    var applicationLaunchURL = urlScheme + "://" + baseURL;
 
-    log.debug("Deployment URL: " + baseURL + deploymentPath);
-    return baseURL + deploymentPath;
+    log.debug("Deployment URL: " + applicationLaunchURL);
```
Comment (lalaji) [inline modules/jaggeryapps/appmgt/src/modules/application/application.jag]: Add debug logs if only debug level enabled.

Comment (lmanohara) [inline modules/jaggeryapps/appmgt/src/modules/application/application.jag]: This is fixed in following commit.
https://github.com/wso2/app-cloud/pull/262/commits/c758e43c2b5c6bd2484a7b8d5689088906dee4d7

**Pull request merged**